chainer/_link.py

```diff
@@ -0,0 +1,373 @@
+import sys
+import typing as tp  # NOQA
+import warnings
+
+import numpy
+
+import chainer
+from chainer import backend
+from chainer.backends import _cpu
+from chainer.backends import cuda
+from chainer.backends import intel64
+from chainer import types  # NOQA
+import chainerx
+
+
+class DeviceResidentMixin(object):
+    """A mix-in class to support multi-device hierarchy.
+    """
+
+    _device = _cpu.CpuDevice()
+
+    # Indicates the object is mixed-device.
+    _MIXED_DEVICE = object()
+
+    def __init__(self):
```

Comment:
Can we check the overrides at the inheritance time instead of instantiation time? Maybe it needs metaclass.

Reply:
I just wrote a simple implementation for such a check. It should be simple to add warning version too. It may be good to put this kind of things to utils. https://gist.github.com/beam2d/3d1cd3baca30793ff804e7075051d6da

Reply:
Hmm, it seems it causes metaclass conflict at the declaration of `Chain`:
```
chainer/link.py:998: in <module>
    class ChainList(Link, collections_abc.MutableSequence):
E   TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

Reply:
It looks `collections_abc.MutableSequence` is a metaclass of type `abc.ABCMeta`, so making `enable_final` a subclass of `ABCMeta` resolves the conflict.

Reply:
Fixed as suggested.
Interface has been slightly changed from the original snippet, so please take a look.